WordPress/gutenberg
# Try: bundle `WP_Theme_JSON` class instead of inheriting per WordPress version

## What?

This PR proposes a different way to organize theme.json code.

## Why?

The existing mechanism uses an inheritance model, where Gutenberg modifies only certain parts of WordPress. This approach:

- has proven buggy: we've missed many backports.
- is difficult to maintain: syncing core<=>Gutenberg is time-consuming and error-prone.
- does not allow us to do certain things, for example, _remove_ specific methods, change the signature, etc.
- we've done some hacks to clean the static variables set by core: [42756](https://github.com/WordPress/gutenberg/pull/42756) and [42776](https://github.com/WordPress/gutenberg/pull/42776).

## How?

The current mechanism relies on inheritance. The code in Gutenberg inherits from the core class and modifies the relevant parts of it, scoped to a specific WordPress version so backports were easier. This means that we have:

- The `WP_Theme_JSON` class that lives in the runtime.
- The `WP_Theme_JSON_6_1` class at `lib/compat/wordpress-6.1` for features to be shipped in WordPress 6.1.
- The `WP_Theme_JSON_6_2` class at `lib/compat/wordpress-6.2` for features to be shipped in WordPress 6.2.
- The `WP_Theme_JSON_Gutenberg` class at `lib/compat/experimental` for features that are not ready to be shipped and/or as a placeholder. The whole Gutenberg codebase we should use `WP_Theme_JSON_Gutenberg` to have the latest changes.

This PR proposes that we have a single class, `WP_Theme_JSON_Gutenberg`, that does not inherit from the core one. Comparing the core and the Gutenberg classes is a matter of using a diff tool over a single file.

## Questions

**Is there a better setup?**

I don't think the current setup is working for the reasons I mentioned above, so it's best if we change it. Happy to hear alternatives if people have them.

**Where this single class should live?**

Initially, this PR proposed to have it at `lib/compat/wordpress-6.2`. It was switched to live at `lib/`, which follows what we do for `lib/block-supports` and other files (`client-assets.php`).

**Using the `@since` tag**

The expectations for this class going forward should be that we use `@since` PHPDoc tag to document the changes like we do in WordPress core.

## Testing Instructions

Given this moves code from one place to another, it's difficult to provide a test plan. The fact that automated tests pass gives us some certainty the main paths are working.

This is what I've done, though you may want to try things you're most familiar with:

Theme with `theme.json` (TwentyTwentyThree):

- In the site editor, make some changes to the global styles (e.g.: select a different style variation) and verify they are applied everywhere (editor, front-end).
- In the site editor, I've added a new color to the global palette and updated an existing one.
- In the editors, verify that presets (typography, colors, etc.) are properly listed (no values missing, etc.).

Theme without `theme.json` (TwentyTwenty):

- Load the theme in the editor and front-end. Verify that it works as before.
- In the editors, verify that presets (typography, colors, etc.) are properly listed (no values missing, etc.).

## Next steps

I'd like to merge this one as soon as the tests pass. This is an active area of development and it the PR is up for more than a few days is going to consume a lot of time to keep it up with the changes.

The same should be done for `WP_Theme_JSON_Resolver`. Given this is a big change in terms of lines of code touched, I'd prefer to do it individually, to manage scope, if at all possible.
